src/Modules/LearningTasks/Tutor.LearningTasks.Core/UseCases/Learning/TaskProgressService.cs

```diff
@@ -66,9 +75,89 @@ public Result<TaskProgressDto> SubmitAnswer(int unitId, int id, StepProgressDto
 
         var taskProgress = result.Value;
         taskProgress.SubmitAnswer(stepProgress.StepId, stepProgress.Answer!);
+        _progressRepository.UpdateEvents(taskProgress);
+
         return Update(taskProgress);
     }
 
+    public Result OpenSubmission(int unitId, int id, int stepId, int learnerId)
```

Comment:
I think all of these non-state changing methods can be simplified to reduce code duplication. This might not work, as I am writing from memory instead of in an IDE:
```csharp
public Result PlayExampleVideo(int unitId, int id, int stepId, int learnerId)
{
    return HandleNonStateChangingEvent(unitId, id, stepId, learnerId, (taskProgress, step) => taskProgress.PlayExampleVideo(step));
}

public Result PauseExampleVideo(int unitId, int id, int stepId, int learnerId)
{
    return HandleNonStateChangingEvent(unitId, id, stepId, learnerId, (taskProgress, step) => taskProgress.PauseExampleVideo(step));
}

public Result FinishExampleVideo(int unitId, int id, int stepId, int learnerId)
{
    return HandleNonStateChangingEvent(unitId, id, stepId, learnerId, (taskProgress, step) => taskProgress.FinishExampleVideo(step));
}

private Result HandleNonStateChangingEvent(int unitId, int id, int stepId, int learnerId, Action<TaskProgress, int> action)
{
    var result = GetTaskProgress(unitId, learnerId, id);
    if (result.IsFailed)
        return result.ToResult();

    var taskProgress = result.Value;
    action(taskProgress, stepId);
    _progressRepository.UpdateEvents(taskProgress);

    return UnitOfWork.Save();
}
```

Reply:
Nice